Python/Product/Analysis/LanguageServer/Hooks.cs

```diff
@@ -0,0 +1,27 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using Microsoft.PythonTools.Parsing.Ast;
+
+namespace Microsoft.PythonTools.Analysis.LanguageServer.Hooks {
+    public sealed class CompletionEventArgs : EventArgs {
+        public ModuleAnalysis Analysis;
```

Comment:
I'd make it with readonly get properties

Reply:
I didn't want to write a constructor, and some of these need to be fields because they're value types.

Reply:
This instance is shared between all subscribers. One subscriber shouldn't be able to affect the next one by changing args unless it is intentionally expected.

Reply:
Also, if we want to be sure that extension has decided if it needs this notification during registration, it is better to provide it with interface it can implement.

Reply:
I don't want to provide that many interfaces (one per overridable call). And your point is well taken on it being shared - I'll make the input parameters readonly properties and leave the return value as a mutable field.